Model Spring `util`

sauyon opened the pull request.

joefarebrother commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+        "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapKey of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[1];MapValue of Argument[-1];value",
```
Would it make sense to use `Element of MapValue` throughout? Since `MultiValueMap<K,V>` extends `Map<K,List<V>>`, it could be used as one.

aschackmull replied:
That's a good observation. Indeed that means that we should use `Element of MapValue` as you suggest.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+  private class FlowSummaries extends SummaryModelCsv {
+    override predicate row(string row) {
+      row = [
+        "org.springframework.util;AutoPopulatingList;false;AutoPopulatingList;(java.util.List<E>, org.springframework.util.AutoPopulatingList$ElementFactory<E>);;Element of Argument[0];Element of ReturnValue;value",
+        "org.springframework.util;AutoPopulatingList;false;AutoPopulatingList;(java.util.List<E>, Class<? extends E>);;Element of Argument[0];Element of ReturnValue;value",
```
Signatures won't match. Spaces aren't allowed, and you should use the erased types.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+        "org.springframework.util;Base64Utils;false;encodeToString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;Base64Utils;false;encodeToUrlSafeString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;Base64Utils;false;encodeUrlSafe;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;CollectionUtils;false;arrayToList;;;Element of Argument[0];Element of ReturnValue;value",
```
Looks wrong: `ArrayElement of` ?

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+        "org.springframework.util;CollectionUtils;false;findValueOfType;;;Element of Argument[0];Element of ReturnValue;value",
+        "org.springframework.util;CollectionUtils;false;firstElement;;;Element of Argument[0];Element of ReturnValue;value",
+        "org.springframework.util;CollectionUtils;false;lastElement;;;Element of Argument[0];Element of ReturnValue;value",
+        "org.springframework.util;CollectionUtils;false;mergeArrayIntoCollection;;;Element of Argument[0];Element of Argument[1];value",
```
`ArrayElement`?

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+        "org.springframework.util;CollectionUtils;false;mergeArrayIntoCollection;;;Element of Argument[0];Element of Argument[1];value",
+        "org.springframework.util;CollectionUtils;false;mergePropertiesIntoMap;;;MapKey of Argument[0];MapKey of Argument[1];value",
+        "org.springframework.util;CollectionUtils;false;mergePropertiesIntoMap;;;MapValue of Argument[0];MapValue of Argument[1];value",
+        "org.springframework.util;CollectionUtils;false;toArray;;;Element of Argument[0];Element of ReturnValue;value",
+        "org.springframework.util;CollectionUtils;false;toArray;;;Element of Argument[0];Element of Argument[1];value",
```
`ArrayElement` ?

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+        "org.springframework.util;FileCopyUtils;false;copyToByteArray;;;Argument[0];Argument[-1];taint",
+        "org.springframework.util;FileCopyUtils;false;copyToString;;;Argument[0];Argument[-1];taint",
+        "org.springframework.util;FileSystemUtils;false;copyRecursively;;;Argument[0];Argument[-1];taint",
+        "org.springframework.util;LinkedMultiValueMap;false;LinkedMultiValueMap;(Map<K,List<V>>);;MapKey of Argument[0];MapKey of ReturnValue;value",
+        "org.springframework.util;LinkedMultiValueMap;false;LinkedMultiValueMap;(Map<K,List<V>>);;Element of MapValue of Argument[0];Element of MapValue of ReturnValue;value",
```
Signatures won't match. Use erased types.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[1];Element of MapValue of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;addAll;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;addAll;(K, java.util.List<? extends V>);;Element of Argument[1];Element of MapValue of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;addAll;(MultiValueMap<K,V>);;MapValue of Argument[0];Element of MapValue of Argument[-1];value",
```
same

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,155 @@
+ * Provides models for the `org.springframework.util` package.
+ */
+module SpringUtil {
+  private class FlowSummaries extends SummaryModelCsv {
```
Does this even compile? There's no import of `ExternalFlow.qll`.

sauyon commented:
I believe I've addressed all the review comments, as well as fixing a few errors discovered while generating tests.

I've also modeled `AntPathMatcher`, which I'd originally skipped thinking that the paths wouldn't be relevant, but on second thought I feel that they should be modeled.

The tests for constructors don't appear to be working and I'm currently not sure why.

sauyon commented:
Stubs are currently in the test directory because technically they're stubs of Spring 5.3.8 because one of the classes modeled (`MultiValueMapAdapter`) is not in 5.2.3; I can obviously move those.

smowton commented:
Constructor tests: I believe it's because we're using `Argument[-1]` to refer to the constructor result, not `ReturnValue`

smowton commented:
Stubs: suggest just renaming the existing stub dir to 5.3.8 and allowing people to incrementally bring the stubs forwards as and when particular use cases dictate

sauyon commented:
Ok, this should hopefully be ready to go now.

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/test.ql:
```diff
@@ -0,0 +1,71 @@
+import semmle.code.java.dataflow.ExternalFlow
+import semmle.code.java.dataflow.TaintTracking
+import TestUtilities.InlineExpectationsTest
+import semmle.code.java.frameworks.spring.SpringUtil
```
Remove. `ExternalFlow` needs bidirectional import, so `SpringUtil` should be imported there instead.

sauyon replied:
At the moment it seems like the Spring models are not imported by default, which is why I didn't add the import anywhere---I guess the taint flow steps should always be included?

aschackmull replied:
Yes. Otherwise they don't actually accomplish anything - as you'll see in this test if you remove this line.

sauyon replied:
Right, but that's true of all the Spring models, as far as I can tell? I have included this in `Spring.qll`.

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/test.ql:
```diff
@@ -0,0 +1,71 @@
+    row =
+      [
+        //"package;type;overrides;name;signature;ext;inputspec;outputspec;kind",
+        "generatedtest;Test;false;newWithArrayElement;;;Argument[0];ArrayElement of ReturnValue;value",
+        "generatedtest;Test;false;getArrayElement;;;ArrayElement of Argument[0];ReturnValue;value",
+        "generatedtest;Test;false;getElement;;;Element of Argument[0];ReturnValue;value",
+        "generatedtest;Test;false;newWithElement;;;Argument[0];Element of ReturnValue;value",
+        "generatedtest;Test;false;getMapKey;;;MapKey of Argument[0];ReturnValue;value",
+        "generatedtest;Test;false;newWithMapKey;;;Argument[0];MapKey of ReturnValue;value",
+        "generatedtest;Test;false;newWithMapValue;;;Argument[0];MapValue of ReturnValue;value",
+        "generatedtest;Test;false;getMapValue;;;MapValue of Argument[0];ReturnValue;value"
```
Remove. Replace with proper implementations. See https://github.com/github/codeql/pull/6130

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/springutil.ssv:
```diff
@@ -0,0 +1,144 @@
+org.springframework.util;AntPathMatcher;false;combine;;;Argument[0..1];ReturnValue;taint
```
What is this file? Something committed by mistake?

sauyon replied:
It's the file used to generate the original form of the tests.

sauyon replied:
I've removed it since I didn't use it to regenerate test cases for the changes I made; not sure whether you wanted it in the first place.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,166 @@
+/**
+ * Provides models for the `org.springframework.util` package.
+ */
+module SpringUtil {
```
No need for this module.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,166 @@
+        "org.springframework.util;AntPathMatcher;false;doMatch;;;Argument[0];MapValue of Argument[2];taint",
+        "org.springframework.util;AntPathMatcher;false;extractPathWithinPattern;;;Argument[1];ReturnValue;taint",
+        "org.springframework.util;AntPathMatcher;false;extractUriTemplateVariables;;;Argument[1];MapValue of ReturnValue;taint",
+        // Return values here are arrays, but there's no value in adding `ArrayValue of` for taint, right?
```
Yes, I'd say there is.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,166 @@
+        "org.springframework.util;StringUtils;false;delimitedListToStringArray;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;getFilename;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;getFilenameExtension;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;mergeStringArrays;;;Argument[0..1];ReturnValue;taint",
```
This looks like it's missing some `ArrayElement of`. (I was triggered by the method name - are there others like it?)

sauyon commented:
I've redone a lot of the `StringUtil` models that were done pre-collection modeling and blindly copy-pasted into this PR.

There are two test cases that fail and confuse me greatly.

I noticed there aren't any models of elements of `Optional`; should I just consider `unwrapOptional` to taint its output and not have value flow as it's written now?

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,161 @@
+        "org.springframework.util;Base64Utils;false;encodeToUrlSafeString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;Base64Utils;false;encodeUrlSafe;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;CollectionUtils;false;arrayToList;;;ArrayElement of Argument[0];Element of ReturnValue;value",
+        "org.springframework.util;CollectionUtils;false;findValueOfType;;;Element of Argument[0];ReturnValue;value",
```
Missing model: `findFirstMatch`

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,161 @@
+        "org.springframework.util;CollectionUtils;false;unmodifiableMultiValueMap;;;MapKey of Argument[0];MapKey of ReturnValue;value",
+        "org.springframework.util;CollectionUtils;false;unmodifiableMultiValueMap;;;MapValue of Argument[0];MapValue of ReturnValue;value",
+        "org.springframework.util;CompositeIterator;false;add;;;Element of Argument[0];Element of Argument[-1];value",
+        "org.springframework.util;CompositeIterator;false;next;;;Element of Argument[-1];ReturnValue;value",
```
```suggestion
```
(inherited)

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,161 @@
+        "org.springframework.util;CollectionUtils;false;unmodifiableMultiValueMap;;;MapValue of Argument[0];MapValue of ReturnValue;value",
+        "org.springframework.util;CompositeIterator;false;add;;;Element of Argument[0];Element of Argument[-1];value",
+        "org.springframework.util;CompositeIterator;false;next;;;Element of Argument[-1];ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Reference;false;get;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Reference;false;get;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Reference;false;get;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Segment;false;getReference;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.util.ConcurrentReferenceHashMap;Segment;false;getReference;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
```
```suggestion
        "org.springframework.util;ConcurrentReferenceHashMap$Reference;false;get;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
        "org.springframework.util;ConcurrentReferenceHashMap$Reference;false;get;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
        "org.springframework.util;ConcurrentReferenceHashMap$Segment;false;getReference;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
        "org.springframework.util;ConcurrentReferenceHashMap$Segment;false;getReference;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
```

(missing tests?)

smowton replied:
These are also all protected classes; dropped

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,161 @@
+        "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;add;;;Argument[1];Element of MapValue of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;addAll;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;addAll;(java.lang.Object,java.util.List);;Element of Argument[1];Element of MapValue of Argument[-1];value",
+        "org.springframework.util;MultiValueMap;true;addAll;(org.springframework.util.MultiValueMap);;Element of MapValue of Argument[0];Element of MapValue of Argument[-1];value",
```
```suggestion
        "org.springframework.util;MultiValueMap;true;addAll;(java.lang.Object,java.util.List);;Argument[0];MapKey of Argument[-1];value",
        "org.springframework.util;MultiValueMap;true;addAll;(org.springframework.util.MultiValueMap);;MapKey of Argument[0];MapKey of Argument[-1];value",
        "org.springframework.util;MultiValueMap;true;addAll;(java.lang.Object,java.util.List);;Element of Argument[1];Element of MapValue of Argument[-1];value",
        "org.springframework.util;MultiValueMap;true;addAll;(org.springframework.util.MultiValueMap);;Element of MapValue of Argument[0];Element of MapValue of Argument[-1];value",
```

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,161 @@
+        "org.springframework.util;MultiValueMapAdapter;false;MultiValueMapAdapter;;;Element of MapValue of Argument[0];Element of MapValue of Argument[-1];value",
+        "org.springframework.util;ObjectUtils;false;addObjectToArray;;;ArrayElement of Argument[0];ArrayElement of ReturnValue;value",
+        "org.springframework.util;ObjectUtils;false;addObjectToArray;;;Argument[1];ArrayElement of ReturnValue;value",
+        "org.springframework.util;ObjectUtils;false;getDisplayString;;;Argument[0];ReturnValue;taint",
```
```suggestion
```

`toString` doesn't convey taint by default, so this shouldn't either

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,161 @@
+        "org.springframework.util;ObjectUtils;false;addObjectToArray;;;ArrayElement of Argument[0];ArrayElement of ReturnValue;value",
+        "org.springframework.util;ObjectUtils;false;addObjectToArray;;;Argument[1];ArrayElement of ReturnValue;value",
+        "org.springframework.util;ObjectUtils;false;getDisplayString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;ObjectUtils;false;identityToString;;;Argument[0];ReturnValue;taint",
```
```suggestion
```

Similarly this shouldn't per default convey taint, nor should `nullSafeToString`

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,161 @@
+        "org.springframework.util;AntPathMatcher;false;extractUriTemplateVariables;;;Argument[1];MapValue of ReturnValue;taint",
+        "org.springframework.util;AntPathMatcher;false;tokenizePath;;;Argument[0];ArrayValue of ReturnValue;taint",
+        "org.springframework.util;AntPathMatcher;false;tokenizePattern;;;Argument[0];ArrayValue of ReturnValue;taint",
+        "org.springframework.util.AntPathMatcher;AntPathStringMatcher;false;matchStrings;;;Argument[0];MapValue of Argument[1];taint",
```
Actually turns out the class is protected so dropping this line

smowton commented:
Pushed fixes for my comments above; @aschackmull I think this should be ready to go

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/test.ql:
```diff
@@ -0,0 +1,54 @@
+import semmle.code.java.dataflow.ExternalFlow
+import semmle.code.java.dataflow.TaintTracking
+import TestUtilities.InlineExpectationsTest
+import semmle.code.java.frameworks.spring.Spring
```
Remove. If the test requires this to pass then it just shows that bidirectional import is missing.
```suggestion
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/test.ql:
```diff
@@ -0,0 +1,53 @@
+import java
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.ExternalFlow
```
This import looks unused.
```suggestion
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+// Test case generated by GenerateFlowTestCase.ql
+public class Test {
+
+	Object getArrayElement(Object container) { return ((Object[])container)[0]; }
```
```suggestion
	Object getArrayElement(Object[] container) { return container[0]; }
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+	Object getMapKey(Map container) { return container.keySet().iterator().next(); }
+	Object getMapValue(Map container) { return container.get(null); }
+	Object[] newWithArrayElement(Object element) { return new Object[] {element}; }
+	// Object newWithMapKey(Object element) { return null; }
```
```suggestion
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+	Object getMapValue(Map container) { return container.get(null); }
+	Object[] newWithArrayElement(Object element) { return new Object[] {element}; }
+	// Object newWithMapKey(Object element) { return null; }
+  Properties newPropertiesWithMapKey(Object element) { Properties p = new Properties(); p.put(element, null); return p; }
```
Some indentation gone wrong here.

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+	Object[] newWithArrayElement(Object element) { return new Object[] {element}; }
+	// Object newWithMapKey(Object element) { return null; }
+  Properties newPropertiesWithMapKey(Object element) { Properties p = new Properties(); p.put(element, null); return p; }
+	// Object newWithMapValue(Object element) { return null; }
```
```suggestion
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+		{
+			// "org.springframework.util;CollectionUtils;false;arrayToList;;;ArrayElement of Argument[0];Element of ReturnValue;value"
+			List out = null;
+			Object in = (Object)newWithArrayElement(source());
```
```suggestion
			Object[] in = newWithArrayElement(source());
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+		{
+			// "org.springframework.util;CollectionUtils;false;mergeArrayIntoCollection;;;ArrayElement of Argument[0];Element of Argument[1];value"
+			Collection out = null;
+			Object in = (Object)newWithArrayElement(source());
```
```suggestion
			Object[] in = newWithArrayElement(source());
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+			MultiValueMap out = null;
+			Map in = Map.of(null, List.of(source()));
+			out = CollectionUtils.toMultiValueMap(in);
+			sink(getElement((List)getMapValue(out))); // $hasValueFlow
```
If you improve the type of `getMapValue` then this cast can be removed.
```
<K,V> V getMapValue(Map<K,V> container)
```

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+			// "org.springframework.util;CollectionUtils;false;unmodifiableMultiValueMap;;;MapValue of Argument[0];MapValue of ReturnValue;value"
+			MultiValueMap out = null;
+			MultiValueMap in = new LinkedMultiValueMap();
+      in.put(null, source());
```
indentation

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+			// "org.springframework.util;CompositeIterator;false;next;;;Element of Argument[-1];ReturnValue;value"
+			Object out = null;
+			CompositeIterator in = new CompositeIterator();
+      in.add(List.of(source()).iterator());
```
indentation

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+			LinkedMultiValueMap out = null;
+			Map in = Map.of(null, List.of(source()));
+			out = new LinkedMultiValueMap(in);
+			sink(getElement((List)getMapValue(out))); // $hasValueFlow
```
Remove cast

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+		{
+			// "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapKey of Argument[-1];MapValue of ReturnValue;value"
+			LinkedMultiValueMap out = null;
+			LinkedMultiValueMap in = (LinkedMultiValueMap)Map.of(source(), null);
```
This is slightly cheating, but type pruning won't be able to catch this at the moment.

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+		{
+			// "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value"
+			MultiValueMap out = null;
+			Object in = (Object)source();
```
Casting to `Object` is superfluous.
```suggestion
			Object in = source();
```

aschackmull replied:
(More instances of this below)

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+			MultiValueMap out = null;
+			Object in = (Object)source();
+			out.add((Object)null, in);
+			sink(getElement((List)getMapValue(out))); // $hasValueFlow
```
All of these casts to `List` can be removed with the above sketched improvement to `getMapValue`.

aschackmull commented on java/ql/test/library-tests/frameworks/spring/util/Test.java:
```diff
@@ -0,0 +1,1445 @@
+			out = in.getFirst((Object)null);
+			sink(out); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;set;;;Argument[0];MapKey of Argument[-1];value"
+			MultiValueMapAdapter out = null;
+			Object in = (Object)source();
+			out.set(in, null);
+			sink(getMapKey(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;set;;;Argument[0];MapKey of Argument[-1];value"
+			MultiValueMap out = null;
+			Object in = (Object)source();
+			out.set(in, null);
+			sink(getMapKey(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;set;;;Argument[0];MapKey of Argument[-1];value"
+			MultiValueMap out = null;
+			Object in = (Object)source();
+			out.set(in, (Object)null);
+			sink(getMapKey(out)); // $hasValueFlow
```
These three look like duplicates of each other?

aschackmull replied:
@smowton Is this something that can be traced back to a bug in the test generation?

smowton replied:
Two differ (the top one is checking inheritance by `MultiValueMapAdapter`); dropped the last one.

smowton commented:
@aschackmull all fixed; some of my commits overlap with your reviews in time; first new commit is "Import SpringUtil from ExternalFlow.qll"

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringUtil.qll:
```diff
@@ -0,0 +1,153 @@
+ */
+
+import java
+import semmle.code.java.dataflow.ExternalFlow
```
```suggestion
private import semmle.code.java.dataflow.ExternalFlow
```

sauyon commented on java/ql/src/semmle/code/java/frameworks/spring/Spring.qll:
```diff
@@ -32,6 +32,7 @@ import semmle.code.java.frameworks.spring.SpringQualifier
 import semmle.code.java.frameworks.spring.SpringRef
 import semmle.code.java.frameworks.spring.SpringReplacedMethod
 import semmle.code.java.frameworks.spring.SpringSet
+import semmle.code.java.frameworks.spring.SpringUtil
```
Should this import be removed?

aschackmull replied:
I think it makes sense to keep. `Spring.qll` just aims to bulk-import all of the spring-related libraries and even though `SpringUtil` currently doesn't contain any public declarations, meaning that this import doesn't do anything, then any future public additions in `SpringUtil` ought to be imported by `Spring.qll`.
